x-pack/test/login_selector_api_integration/apis/login_selector.ts

```diff
@@ -0,0 +1,505 @@
+          );
+
+          const cookies = authenticationResponse.headers['set-cookie'];
+          expect(cookies).to.have.length(1);
+
+          await checkSessionCookie(request.cookie(cookies[0])!, 'a@b.c', providerName);
+        }
+      });
+
+      // Ideally we should be able to abandon intermediate session and let user log in, but for the
+      // time being we cannot distinguish errors coming from Elasticsearch for the case when SAML
+      // response just doesn't correspond to request ID we have in intermediate cookie and the case
+      // when something else has happened.
+      it('should be able to log in via SP initiated login even if intermediate session with other SAML provider exists', async () => {
+        // First start authentication flow with `saml1`.
+        const saml1HandshakeResponse = await supertest
+          .post('/internal/security/login_with')
+          .ca(CA_CERT)
+          .set('kbn-xsrf', 'xxx')
+          .send({
+            providerType: 'saml',
+            providerName: 'saml1',
+            currentURL: 'https://kibana.com/login?next=/abc/xyz/handshake?one=two%20three#/saml1',
+          })
+          .expect(200);
+
+        expect(
+          saml1HandshakeResponse.body.location.startsWith(`https://elastic.co/sso/saml`)
+        ).to.be(true);
+
+        const saml1HandshakeCookie = request.cookie(
+          saml1HandshakeResponse.headers['set-cookie'][0]
+        )!;
+
+        // And now try to login with `saml2`.
+        const saml2HandshakeResponse = await supertest
+          .post('/internal/security/login_with')
+          .ca(CA_CERT)
+          .set('kbn-xsrf', 'xxx')
+          .set('Cookie', saml1HandshakeCookie.cookieString())
+          .send({
+            providerType: 'saml',
+            providerName: 'saml2',
+            currentURL: 'https://kibana.com/login?next=/abc/xyz/handshake?one=two%20three#/saml2',
+          })
+          .expect(200);
+
+        expect(
+          saml2HandshakeResponse.body.location.startsWith(`https://elastic.co/sso/saml`)
+        ).to.be(true);
+
+        const saml2HandshakeCookie = request.cookie(
+          saml2HandshakeResponse.headers['set-cookie'][0]
+        )!;
+
+        const saml2AuthenticationResponse = await supertest
+          .post('/api/security/saml/callback')
+          .ca(CA_CERT)
+          .set('Cookie', saml2HandshakeCookie.cookieString())
+          .send({
+            SAMLResponse: await createSAMLResponse({ issuer: `http://www.elastic.co/saml2` }),
+          })
+          .expect(302);
+
+        expect(saml2AuthenticationResponse.headers.location).to.be(
+          '/abc/xyz/handshake?one=two three#/saml2'
+        );
+
+        const saml2SessionCookie = request.cookie(
+          saml2AuthenticationResponse.headers['set-cookie'][0]
+        )!;
+        await checkSessionCookie(saml2SessionCookie, 'a@b.c', 'saml2');
+      });
+    });
+
+    describe('Kerberos', () => {
```

Comment:
**question** Since Kerberos and PKI are both handled "transparently" from the end-users perspective, do you think it's worthwhile to add a test making sure that we can authenticate via Kerberos, even when a valid client cert is presented to Kibana?

Reply:
Yeah, it's a good idea - will add!